the-collab-lab/tcl-30-whats-near-me
# [TCL30-16] Allow users to delay API calls

## Description

This PR allows users to set how frequently the app fetches new places. The functionality is necessary to help conserve user device’s battery. The PR presents the user with a slider. Interacting with it allows to set the desired API call frequency and how often the app needs to update. The PR makes sure the functionality is applied only when the user is sharing their location.

## Related Issue

Closes [TCL30-16](https://the-collab-lab.atlassian.net/browse/TCL30-16)

## Acceptance Criteria

- [x] On the settings view, user is presented with a slider as a way to set api calls delay
- [x] The delay should only apply to when the user is sharing their location and their location is the map center. 
- [x] If the user is panning the map manually, there should  be no added delay to the API call
- [x] The delay should be set in seconds with the following values: ``0, 5, 10, 15, 20, 25, 30``
- [x]  The app should wait the selected number of seconds to make each API call

## Type of Changes



|     | Type                       |
| --- | -------------------------- |
|     | :bug: Bug fix              |
| ✓   | :sparkles: New feature     |
|     | :hammer: Refactoring       |
|     | :100: Add tests            |
| ✓   | :link: Update dependencies |
|     | :scroll: Docs              |

## Updates

### Before
Before, the ``Settings`` view had no slider, and the user sharing their location couldn’t set the delay between API calls

Screenshot of `/Settings`

<img width="352" alt="TCL30-16BeforeScreenshot" src="https://user-images.githubusercontent.com/38255956/129512902-3ec46127-c3d9-43a6-9ffc-ab437f23c1e4.png">

### After
Now, in the Settings view, the user is presented with a slider. If their location is shared, they can set the frequency with which the app fetches new places and, thus, conserve their device battery.

[TCL30-16AfterScreenRecordingNetwork760.mov.zip](https://github.com/the-collab-lab/tcl-30-whats-near-me/files/6989961/TCL30-16AfterScreenRecordingNetwork760.mov.zip)

[Video](https://www.loom.com/share/c34b4481fdb9423884cd0433862ad366)  of how app works when user selects 30 seconds, shares their location, and then waits for the API call to render the nearby locations to the user. 




## Testing Steps / QA Criteria

1. In your terminal, from ``tcl30-whats-near-me``, pull down this branch ``git checkout ku-jw-delay-api-calls``
2. Run``npm install`` and ``npm start``  to launch the app 
3. Navigate to the ``Settings view``to make sure it has a slider as a way to set api calls delay
4. For testing purposes, it is suggested to select a longer frequency, such as `20`, `25`, or `30` for the initial test
5. Click `Share My Location` button and **start moving , i.e. changing your position coordinates on the map**
6. Make sure that the app waits the selected number of seconds to make each API call 
7. To validate the correct delay is being applied, open the ``Network`` tab in the browser’s developer tools and inspect how frequently API calls are being made. Once the time set in ``Settings`` has lapsed, you will see locations rendered on the map
8. Note that if you are in an area that might not have many locations that will render, you may zoom out while waiting
9. Make sure the delay only applies to when you are sharing your location and you’re on the move. If you're panning the map manually, there should be no added delay to the API call
.
